src/Interpreters/CollectJoinOnKeysVisitor.cpp

```diff
@@ -668,51 +676,68 @@ void CollectJoinOnKeysMatcher::handleLagBehind(const ASTFunction & func, const A
-            lag_interval_unit = 1;
-        else if (interval_func->name == "to_interval_second")
-            lag_interval_unit = 1000;
-        else if (interval_func->name == "to_interval_minute")
-            lag_interval_unit = 60 * 1000;
+        Int64 lag_interval_unit = 0;
+        if (auto * interval_func = func.arguments->children.at(0)->as<ASTFunction>())
+        {
+            if (interval_func->name == "to_interval_millisecond")
+                lag_interval_unit = 1;
+            else if (interval_func->name == "to_interval_second")
+                lag_interval_unit = 1000;
+            else if (interval_func->name == "to_interval_minute")
+                lag_interval_unit = 60 * 1000;
+            else
+                throw Exception(
+                    ErrorCodes::NOT_IMPLEMENTED,
+                    "First argument of function {} only supports millsecond, second or minute time interval (examples, 1ms, 1s, 2m etc), but got '{}' instead",
+                    func.name, func.formatForErrorMessage());
+
+            auto * interval = interval_func->arguments->children.at(0)->as<ASTLiteral>();
+            assert(interval);
+            lag_interval = interval->value.get<Int64>() * lag_interval_unit;
+            if (lag_interval <= 0)
+                throw Exception(
+                    ErrorCodes::BAD_QUERY_PARAMETER,
+                    "First argument of function {} shall be positive millisecond, second or minute time interval (examples, 1ms, 2s, 3m etc), but got '{}' instead",
+                    func.name, func.formatForErrorMessage());
+        }
         else
-            throw Exception(
-                ErrorCodes::NOT_IMPLEMENTED,
-                "First argument of function {} only supports millsecond, second or minute time interval (examples, 1ms, 1s, 2m etc), but got '{}' instead",
-                func.name, func.formatForErrorMessage());
+            throw Exception(ErrorCodes::SYNTAX_ERROR, "First argument of function {} shall be a time interval (examples, 1ms, 2s, 3m etc), but got '{}' instead", func.name, func.formatForErrorMessage());
 
-        auto * interval = interval_func->arguments->children.at(0)->as<ASTLiteral>();
-        assert(interval);
-        lag_interval = interval->value.get<Int64>() * lag_interval_unit;
-        if (lag_interval <= 0)
-            throw Exception(
-                ErrorCodes::BAD_QUERY_PARAMETER,
-                "First argument of function {} shall be positive millisecond, second or minute time interval (examples, 1ms, 2s, 3m etc), but got '{}' instead",
-                func.name, func.formatForErrorMessage());
+        data.addLagBehindKeys(lag_interval);
     }
     else
-        throw Exception(ErrorCodes::SYNTAX_ERROR, "First argument of function {} shall be a time interval (examples, 1ms, 2s, 3m etc), but got '{}' instead", func.name, func.formatForErrorMessage());
+    {
+        /// 2) lag_behind(2, col1, col2): the lag unit is based on alignment keys.
```

Comment:
We still need `lag_behind(2ms, ts1, ts2)` as mentioned in other place. And for timestamp, the first argument being always an interval seems very intuitive.